### PR TITLE
Cubed-sphere Jacobian and wind interpolation

### PR DESCRIPTION
**Pull request by odlomax**
This PR implements the `Projection::jacobian` method for the equiangular cubed-sphere. There are three versions of this function:

* `projection::Jacobian Proection::jacobian(const PointXY& xy)` This returns the Jacobian ∂(x, y) / ∂(lon, lat).
* `projection::JacobianProection::jacobian(const PointXY& xy, idx_t t)` This returns the Jacobian ∂(x, y) / ∂(lon, lat) for tile t. This is needed to perform transforms on halo points.
* `projection::Jacobian Proection::alphabetaJacobian(const Point2& alphabeta, idx_t t)` This returns the Jacobian ∂(alpha, beta) / ∂(lon, lat) for tile t. As above, but rotated into the (alpha, beta) coordinate system.

Two tests have been added:

1. Convergence test, added to `tests/projection/test_cubedsphere_projection.cc`
    This makes sure that xy + J(x, y) * h(lon, lat) converges quadratically as the step size h(lon, lat) tends to zero.
2. A demonstration of how the Jacobian can be used to improve the accuracy of (u, v) wind interpolation, especially at the poles 
    (added to `tests/interpolation/test_interpolation_cubedsphere.cc`). Here, the following operations are performed:
    * An analytic (u, v) field is generated for a C48 cubed sphere grid.
    * The (u, v) is transformed to (v_alpha, v_beta) via the `alphabetaJacobian` method.
    * (v_alpha, v_beta) is interpolated to an O48 Gaussian grid field.
    * The new (v_alpha, v_beta) is transformed back to (u, v) on the new grid.
    * The adjoint of the above is also tested.
    
The figures below show the output u and v field from the wind interpolation test.
u
v

The two figures below show the absolute interpolation error, i.e. the vector distance between (u_interp, v_interp) and (u_analytic, v_analytic). The colour scale is logarithmic.
* Field `error_field_0` shows the error when we naively interpolate u and v as if they were scalar fields. Note the excessive error 
   at the poles (visible near the top of the sphere) where there is a discontinuity in the direction of the (lon, lat) unit vectors. 
error_0
* Field `error_field_1` shows  the error when we perform the above wind transform. Note the absence of the polar error in the previous field. 
error_1

### CODE REVIEW

**Comment by wdeconinck** [comment]
Please integrate the develop branch as PR #98 is merged in now.

**Comment by wdeconinck** [comment]
@odlomax Again a very good contribution and high quality code! 
I'm very pleasantly surprised how the StructuredMeshGenerator just worked with the odd partition shapes. No doubt however some failing corner cases would still pop up when stress-testing.
Are you happy for me to do a squash merge?

**Comment by odlomax** [comment]
> erge?

@wdeconinck Thank you for your kind words, you've been a great help! Squash and merge away! 😀

**Review by wdeconinck — APPROVED** [review]

File: src/atlas/projection/Jacobian.h [hunk 1]
```diff
@@ -82,7 +82,7 @@ class Jacobian : public std::array<std::array<double, 2>, 2> {
         return Jacobian{(*this)[1][1], -(*this)[0][1], -(*this)[1][0], (*this)[0][0]} * (1. / determinant());
     }
 
-    Jacobian transpose() const { return Jacobian{(*this)[1][1], (*this)[0][1], (*this)[1][0], (*this)[0][0]}; }
+    Jacobian transpose() const { return Jacobian{(*this)[0][0], (*this)[1][0], (*this)[0][1], (*this)[1][1]}; }
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/projection/Jacobian.h:85]: Yes, that was a bit of a head-scratcher! 😂

File: src/tests/interpolation/test_interpolation_cubedsphere.cc [hunk 2]
```diff
@@ -125,6 +159,228 @@ CASE("cubedsphere_interpolation") {
...
+        auto v            = array::make_view<double, 1>(sourceFieldSet["v_orig"]);
+        auto vAlpha       = array::make_view<double, 1>(sourceFieldSet["v_alpha"]);
+        auto vBeta        = array::make_view<double, 1>(sourceFieldSet["v_beta"]);
+
+
+        // In practice, the (u, v) field needs to be halo exchanged *before* the
+        // wind transform. Then the transform is applied to the entire field,
+        // *including* the halo.
+
+        sourceFunctionspace.parallel_for(util::Config("include_halo", true), [&](idx_t idx, idx_t t, idx_t i, idx_t j) {
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/tests/interpolation/test_interpolation_cubedsphere.cc:208]: According @MarekWlasak I'm obsessed with `std::tie`. I think he's right! 😂

**Comment by wdeconinck** [comment]
Squash-merged in develop with commit https://github.com/ecmwf/atlas/commit/1b08c233c45faae9ab846e51da8376057b72e5fa

**Pull request closed without merging**